### DISCUSSION TITLE
If [Window] and [Website]?

### DISCUSSION THREAD

**Discussion by SuitYourself** [WarmUpTill/SceneSwitcher#631] [Q&A] [answered]
Hi there, brilliant addon.

I was wondering if there is anyway that we could check the website the browser is on?

For example, I have a scene for YouTube with my youtube details, I have a scene for Rumble with my rumble details, Twitch with my twitch details etc and so on.  All colour co-ordinated etc. 

It would be nice if I go to youtube in the browser and no matter what sub section of youtube or youtube video I am playing, the youtube obs scene would be on.  If I switch to twitch and either stream my own page or another persons, my obs twitch theme scene would come up.

Essentially, I need a check that would check "youtube.com/*.*" and "rumble.com/*.*" and "twitch.tv/*.*"

The above would be brilliant. Currently, unfortunately I have found that if my Twitch window name is Suit and my YouTube window name is Suit and my Rumble window name is Suit, well, only the first macro 'Suit' match will trigger, and thus if that is for Rumble, then it will show the rumble theme as I browse youtube or twitch. 

Is there any possible way you could add a check like mentioned about? Even if it is some kind of variation of the GET http option to check what the website origin is?

Thank you so much for reading and I do hope it is possible and you get back to me.  I understand, this time of year is a busy one, so if I don't hear back beforehand, have a great christmas and a wonderful, safe New Year!

Kind regards,
Suit.

**Comment by WarmUpTill** [comment]
Hey,

>I was wondering if there is anyway that we could check the website the browser is on?

Just to clarify:
You're referring to a browser application like Firefox, Chrome, etc. and not the "Browser" source type that is available in OBS, are you not?

If so I cannot really think of a good way to do this, if the window title of the particular browser application is always the same on these different websites.
Maybe extensions or settings are available to allow you to customize the window title depending on which URL you are currently visiting?

Directly interacting with browsers might prove to be difficult.

Alternatively - although admittedly a bit resource intensive for such a simple task - you could try to match certain image patterns using the "Video" condition type to identify each website.

**Comment by SuitYourself** [comment]
Hi there, yeah that is correct. Web Browser like Chome Brave Etc.

I'll check in to browser extensions to see if it is possible.

Image patterns... interesting.  I'll look in to that and see if I can pull something off.  It is such a shame we can't get a tab title's url.  I'll play around with the video image matching you mentioned and see what I can get happening.

Thanks for the prompt reply, WarmUpTill.

**Comment by SuitYourself** [comment]
So, I have made all the macros an image search now, and it all works (having to wait for it to do so, so not an instant check on click).
 
Which brings me to my next question/suggestion: Any way to register a left click or add an option under the image search options [Perform check on left mouse click]

Thanks, WarmUpTill.  Appreciate your addon and help. :)

Ideally would be great to have something like: 
If [process] Brave/Chrome is in [Focus] 
and [Mouse LeftClick] = True or [Enter Key] = pressed, 
Then do [Video] Image Search.

**Comment by SuitYourself** [comment]
I think there is a bug...

If I am understanding this correctly the top window is the IF statements. IF X is TRUE then do the bottom window outcome.
The bottom window is the outcome window, if the top window conditions are true then do the conditions within.

So in my example, top window is match video image capture, capturing a part of the screen shown and if it matches, then we're a go to do the bottom which would be switch scene to xyz.

Problem is, the bottom condition is running regardless of the matched image.
![image](https://user-images.githubusercontent.com/25052021/208294004-c3382514-3e92-49f6-8912-e3a8d9beef71.png)

In this example posted, the bottom scene switcher is running although the top condition is not running as it hasn't matched the image as I am on a screen that doesn't have the logo image I asked it to match.

Is this a bug cause it is doing my head in.  It has to be a bug, I can't find any other way to stop it from running the scene switcher in the bottom area if top area is not a match... which it should do right?

Each of the mediums listed on the left side from youtube down to minds all have their own individual unique logo image cut and pasted into a file and added for each one to search, so they are all different.

I make it so that none of these run until the mouse area macro detects the mouse in that top area of the web browser (since we don't have a mouse click option (yet?)).

![image](https://user-images.githubusercontent.com/25052021/208294668-e9ef7d31-bbaa-40d0-8a2d-a299f6450f11.png)

So if mouse cursor in that area and moves, then run each of the macros listed one after another.
This technically in my mind should be find since only one of them should technically produce a match and switch scenes. But they all run the scene switcher regardless of the image match.

Did I do something wrong? Or did I find a bug? :)

Thanks for reading!
Kind regards,
Mark.

PS: In testing these, I presume also I should just be able to click the 'run macro' button and it should bypass the bottom window outcomes if the top window conditions are not met.  Well, when I click run macro, the bottom outcome runs regardless of the conditions above.

PSS: Note the viceo window being used is its own window which I have moved out of sight and off the obs visual area. This doesn't affect the matching of the image though, each macro correctly finds the image in the mentioned window.

PSSS: The only thing I can logically think is that it is detecting an image in one of the macros to be TRUE, and it is then throwing this value to each other macro being run, regardless of it being an incorrect image, the originally detected image was true. 

IF the above is the case, then we need some way to clear and wipe this image value being TRUE upon a new macro being run, or at the end of the scene switch being made so that the next macro in the chain doesn't have it set to TRUE.

**Comment by SuitYourself** [comment]
Hmm... I even attempted to create my own variable called Medium Watch, which is set to NONE.
So in my medium list, I have a if variable Medium Watch is NONE then do the image search on the top window...
And on the bottom window, after switching scenes, I changed the variable Medium Watch to in this case, YOUTUBE.

![image](https://user-images.githubusercontent.com/25052021/208295626-81d9a38a-5499-412a-8945-2539ae4a5525.png)

And in the Twitch Macro it should only check if the variable is NONE and if it does check it should change the variable to TWITCH.

![image](https://user-images.githubusercontent.com/25052021/208295707-83a9812b-112d-4549-a59b-ae7cbae1e104.png)

And the only running macro would be my MOUSE AREA macro which if in the correct  area, should reset the variable to NONE then play each medium Macro in order...

![image](https://user-images.githubusercontent.com/25052021/208295766-839ec65e-f499-420f-80ef-8601589d862a.png)

So my logic is, the variable is changed by the correct matching medium image obtained and thus shouldn't run on any other medium Macro as it is no longer set to NONE... 

of course I shouldn't need to add this check as I think there may be a bug as mentioned in the previous comment however, I figured I would attempt to bypass it with my own home made variable, but that did not even work.

So... either I'm missing something on how this works, or there is a bug or two I've found. :)
Thanks for reading.
Kind regards,
Mark.

**Comment by nacalai** [comment]
Hi there.  So, I have solution to what you are asking. It seems like you managed to figure everything out with imagesearch etc. I would like to anyway tip you about this other method. Like WarmUpTill said, you will need the help of extensions. We're gonna get help of a chrome plugin called [AutoControl](https://www.autocontrol.app/).

AutoControl is basically an advanced scene switcher, for Chrome. It can read tabs, run scripts and it'll let you automate anything. I'm a big fan of it, and it's a super tool to enable OBS with scene switcher, and Chrome, to basically talk to eachother.

In short terms:

**In AutoControl**
Create a trigger: - If tab focus (or) tab opens (or) tab load (and) contains "youtube" in the url -> run a script that writes a local txt file with the word "youtube" in it.

Then you just duplicate this for twitch and other websites, and use the same txt file.

So now, the txt file will be constantly updated depending on which tab you have open.

![image](https://user-images.githubusercontent.com/4357694/208298340-45086498-2ee4-461b-82f3-9cedd57e40dc.png)


**In Advanced Scene Switcher**
Add a macro that reads the file content. So, if file content = youtube, switch to youtube scene, etc.
![image](https://user-images.githubusercontent.com/4357694/208298373-15caa193-9ef0-4d16-865a-5437ddab8b97.png)

So, now you get the picture :) I believe this to be a quick and easy way to accomplish what you are looking for. Using a txt file is an easy way to establish a connection between OBS and other programs and extensions.

**Comment by SuitYourself** [comment]
Hi there nacalai.  Thank you for the idea, I'll give it a try but I have a feeling that the same bug will be present that I have found listed above.

Thank you very much for your suggestion. Downloading and giving it a try now. :)

**Comment by SuitYourself** [comment]
![image](https://user-images.githubusercontent.com/25052021/208300102-b2ed3d14-e0bd-4096-be64-1aad6c461a4f.png)
Oof, this doesn't look too trustworthy, the Native Component they ''require''.... thoughts?

**Reply by nacalai** [reply]
That's to be expected. The native component is what allows the extension to write files to the pc, run programs etc. So, it's a false positive warning. This plugin would've been shut off and removed from the store ages ago, if it actually did harm

**Comment by SuitYourself** [comment]
Ok, @nacalai I have your browser extension running perfectly, it changes the textfile based on what url is in the tab.
All working flawlessly.

Now, @WarmUpTill and @nacalai When I have separate macros for each, the file is loaded in and read, the correct scene is switched, other macros are ignored as per expected (so one could say this is done and fixed and solved)

![image](https://user-images.githubusercontent.com/25052021/208303281-321a8db3-ce9d-48ba-b011-624379bc52e0.png)

BUT, NO. The bug I found still exists.

![image](https://user-images.githubusercontent.com/25052021/208303323-1b3d482a-6000-4d68-8b85-cde4e2fc2da6.png)

If I untick all other macros and just use my one 'Mouse Area' macro, the bug remains.  

Why would I want one Macro to trigger multiple? Cause I only wanted the macros to trigger when I click on a tab, since we don't have 'on mouse click' options in Advanced Scene Switcher, I went with Mouse Cursor movement in a specfic area of the screen.

THIS is suppose to make it so the macros are not continously running in the backgound taking up resources, instead, only one is being triggered when the mouse enters an area (or hopefully 'on click' later on).

ONCE this is triggered, it should then check each of the Macros I have created and switch screen accordingly.

That said, I think the same issue is happening (as mentioned above), it is remembering that the matched value is true and then just bypassing each of the conditions and running the outcome box. 

The bug only happens if you are using one macro to call multiple macros using a same search/variable/file/value.

If I don't use the 'MOUSE AREA' macro, and just tick all of the other macros, all works perfectly, but it is continuousy checking and running in the background, which is what I was attempting to bypass. :)

I hope this makes sense.
@nacalai thank you for the chrome addon, using that for now in conjunction with Advanced Scene Switcher.
@WarmUpTill thank you for reading, I hope all of the above made sense and is helpful in debugging.  :)

Kind regards,
Mark

**Comment by WarmUpTill** [comment]
### Unexpected macro execution:
I _think_ what is happening is that you are misunderstanding what the "Run" option of the "Macro" action type does.
It simply executes the actions of the specified macro as if you were pressing the "Run macro" button - so it intentionally skips the conditions check and just performs the actions.
(I _assume_ the option to Un-/pause macros would be more fitting here)

It is a bug however that paused macros are executed.
This is already fixed on the master branch but was not yet released.

### Mouse click:

I implemented the option to check for mouse clicks for Windows.
![Cursor](https://user-images.githubusercontent.com/19472752/208305077-e72f2c86-218b-4303-aa67-1a4398bf0630.PNG)

A build with this change should be available here in a few minutes:
https://github.com/WarmUpTill/SceneSwitcher/actions/runs/3725304533

This might help you to simplify your setup further.
Let me know if you encounter any issues!

**Comment by SuitYourself** [comment]
I understand the 'run macro' button at the top window should run the macro with the conditions applied to it. The same logic in my brain I apply to the 'run' option of the 'macro' option within the bottom window.  I just figured it should run the macros in order and not skip any conditions. Seems dumb to skip conditions. Maybe there should be an option 'Run' and 'Skip Conditions/Don't Skip Conditions' ?

The mouse click option is going to be helpful, I appreciate that! Thank you!

I'll let you know of my progress! 

Much appreciated.

**Comment by SuitYourself** [comment]
@WarmUpTill, still waiting on the update with the Mouse Click option. :)
Thanks in advanced!

**Comment by nacalai** [comment]
@SuitYourself Click on the build, scroll down to artifacts and you'll see the downloads for the build

**Comment by SuitYourself** [comment]
Installed and tested. 
The new left click option works, but I have to click a couple of times on the tab for it to register since I have a separate macro for each medium.

Since I don't understand how to work this run pause unpause macro thing you got going, I can't make one macro which checks all others since all conditions are bypassed when running then.

If you got a way to have one macro play multiple in one hit and check all their conditions, then I'd love to know. I can't figure it out.

Thanks again :)

**Comment by WarmUpTill** [comment]
>If you got a way to have one macro play multiple in one hit and check all their conditions, then I'd love to know.

I am not sure I fully understand what you are trying to do.
So I apologize in advance if the following description does not really match what you are trying to achieve:

You have a single macro that you want to use as some sort of "master" condition and only if its condition are true should any of the other macros be executed (assuming their conditions are true as well), correct?

If so you could simply use a setup like this (the used conditions and actions are just examples here):

![M1](https://user-images.githubusercontent.com/19472752/208879916-6a30e4d7-3203-4204-9d0d-1f85d142b891.PNG)
![M2](https://user-images.githubusercontent.com/19472752/208879933-19c25132-3923-4bd3-bec6-96d990b103c9.PNG)
![M3](https://user-images.githubusercontent.com/19472752/208879945-a930e334-9b0c-4f95-ab79-6eff840de6be.PNG)

The macros "Normal macro 1" and "Normal macro 2" can only ever run if "Master macro's" conditions are true also.

Hope that helps!

**Comment by SuitYourself** [comment]
I'll play around with that idea. Thanks again @WarmUpTill :)

**Comment by SuitYourself** [comment]
Thought I would try another way, and not sure if I discovered another bug or not.

Created two macros:
Browser Focus
Browser Unfocus

Both are ticked to run.
In Browser Focus I have: If Process brave.exe is running [x] and in focus

Then

Macro unpause YOUTUBE (my youtube macro)
Macro unpause TWITCH (my twitch macro)

And Browser Unfocus macro has the same but unticked the 'and is focused' option and canged 'unpause' to 'pause'. 

That said, it doesn't seem toggle pause and unpause based on browser focus.
Independantly, each Browser Focus/Browser Unfocus macro works, but if I have both ticked, then for some reason my YouTube Macro and Twitch Macro remain paused.

![image](https://user-images.githubusercontent.com/25052021/208890138-ef7f5f04-0271-48d7-a17a-6d8dc5fe2172.png)
![image](https://user-images.githubusercontent.com/25052021/208891219-7b1ce363-1a09-42af-95dd-164103d24a6e.png)

**Comment by WarmUpTill** [comment]
"BROWSER UNFOCUS" is not the inverse of "BROWSER FOCUS" so _both_ macros will be executed if a process with the name "brave.exe" exists and is focused.
This will result in the "YOUTUBE" and "TWITCH" macro being unpaused by "BROWSER FOCUS" and then being immediately being paused again by "BROWSER UNFOCUS".

What you _probably_ wanted "BROWSER UNFOCUS" to be is "If **not** ... <conditions of BROWSER FOCUS here>".
For this you can simply use the "Macro" condition type and reference the condition state of the "BROWSER FOCUS" macro.

**Comment by SuitYourself** [comment]
Ah I see (sorry for my constant posts from a Noob.  If anything, you'll see how the average dumb dumb reads things).

I'll set out the [if not] functions in a moment. : )

I hope you don't mind my stupid posts updating you of what I am trying.

**Comment by SuitYourself** [comment, marked as answer]
All working brilliantly now mate! 
I have it so it only need to activate and use the macros when I left click on the browser or back in to my game. <3

Thank you so much for your help @WarmUpTill 
What is your paypal for me to throw you a donation mate?

**Comment by SuitYourself** [comment]
@WarmUpTill Not sure if it is possible, but your 'export/import' option to save our work, any chance it could remember the last saved folder it was in?

For me, I save it in my documents/SceneSwitcher/

Every time I click 'export' I have to navigate to the folder location.
Also, you got a paypal so I can send you a donation mate?

**Reply by WarmUpTill** [reply]
>Not sure if it is possible, but your 'export/import' option to save our work, any chance it could remember the last saved folder it was in?

Sure, that should be doable - I will add it to the todo list.

>Also, you got a paypal so I can send you a donation mate?

Thank you very much for the offer.
I really do appreciate it.
I don't want to sound ungrateful, but please donate the money to a charity or OBS instead.

**Comment by SuitYourself** [comment]
Will do. Thanks again @WarmUpTill and Merry Christmas!
